File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int LAN, num;
    cin >> LAN >> num;

    vector<int> LAN_length(LAN);
    for(int i = 0; i < LAN; i++) {
        cin >> LAN_length[i];
    }
    long long start = 1;
    long long end = *max_element(LAN_length.begin(), LAN_length.end());
    while(start <= end) {
        long long mid = (start + end) / 2;
        int cnt = 0;
        for(int i = 0; i < LAN; i++)
            cnt += LAN_length[i] / mid;
        
        if(cnt >= num)
            start = mid + 1;
        else
            end = mid - 1;
    }

    cout << end << endl;

    return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
LAN, num = map(int, input().split())
LAN_length = []
for _ in range(LAN):
    LAN_length.append(int(input()))

start, end = 1, max(LAN_length)
while start <= end:
    mid = (start + end) // 2
    cnt = 0
    for ln in LAN_length:
        cnt += ln // mid
    if cnt >= num: # 목표치보다 더 많은 랜선이 만들어지는 경우
        start = mid + 1
    else: # 목표치보다 더 작은 랜선이 만들어지는 경우
        end = mid - 1
        
print(end)